rust/worker/src/blockstore/arrow_blockfile/block/delta.rs

```diff
@@ -0,0 +1,382 @@
+    pub fn add(&self, key: BlockfileKey, value: Value) {
+        let mut inner = self.inner.write();
+        inner.add(key, value);
+    }
+
+    /// Deletes a key from the block delta.
+    pub fn delete(&self, key: BlockfileKey) {
+        let mut inner = self.inner.write();
+        inner.delete(key);
+    }
+
+    /// Gets the minimum key in the block delta.
+    pub fn get_min_key(&self) -> Option<BlockfileKey> {
+        let inner = self.inner.read();
+        let first_key = inner.new_data.keys().next();
+        first_key.cloned()
+    }
+
+    /// Gets the size of the block delta as it would be in a block. This includes
+    /// the size of the prefix, key, and value data and the size of the offsets
+    /// where applicable. The size is rounded up to the nearest 64 bytes as per
+    /// the arrow specification. When a block delta is converted into a block data
+    /// the same sizing is used to allocate the memory for the block data.
+    pub fn get_size(&self) -> usize {
+        let inner = self.inner.read();
+        inner.get_size(
+            self.source_block.get_key_type(),
+            self.source_block.get_value_type(),
+        )
+    }
+
+    /// Splits the block delta into two block deltas. The split point is the last key
+    /// that pushes the block over the half size.
+    pub fn split(&self, provider: &ArrowBlockProvider) -> (BlockfileKey, BlockDelta) {
+        let new_block = provider.create_block(
+            self.source_block.get_key_type(),
+            self.source_block.get_value_type(),
+        );
+        let mut inner = self.inner.write();
+        let (split_key, new_adds) = inner.split(
+            self.source_block.get_key_type(),
+            self.source_block.get_value_type(),
+        );
+        (
+            split_key,
+            BlockDelta {
+                source_block: new_block,
+                inner: Arc::new(RwLock::new(BlockDeltaInner { new_data: new_adds })),
+            },
+        )
+    }
+
+    fn get_prefix_size(&self) -> usize {
+        let inner = self.inner.read();
+        inner.get_prefix_size()
+    }
+
+    fn get_key_size(&self) -> usize {
+        let inner = self.inner.read();
+        inner.get_key_size()
+    }
+
+    fn get_value_size(&self) -> usize {
+        let inner = self.inner.read();
+        inner.get_value_size()
+    }
+
+    fn get_value_count(&self) -> usize {
+        let inner = self.inner.read();
+        inner.get_value_count()
+    }
+
+    fn len(&self) -> usize {
+        let inner = self.inner.read();
+        inner.new_data.len()
+    }
+}
+
+struct BlockDeltaInner {
+    new_data: BTreeMap<BlockfileKey, Value>,
+}
+
+impl BlockDeltaInner {
+    fn add(&mut self, key: BlockfileKey, value: Value) {
+        self.new_data.insert(key, value);
+    }
+
+    fn delete(&mut self, key: BlockfileKey) {
+        if self.new_data.contains_key(&key) {
+            self.new_data.remove(&key);
+        }
+    }
+
+    fn get_block_size(
```

Comment:
Something about this feels a little funky to me.

I think it's that we pass in most of the data required to compute this (which makes it seem more like a pure function) but also use only two pieces of state from `self`. Bit of a nit, but maybe this should be a private function not in an `impl` and also add parameters for `key_offset_bytes` and `value_offset_bytes`? I'm fine not doing this now, just thinking

Reply:
Yeah sure we can do that, I just bias towards putting everything in the impl for some reason. Its a way of saying "this is a private method contained to the inner workings of this struct". I don't really have a strong aesthetic opinion here.